### ISSUE TITLE
Does changing of states of server instances in a shop instance constitute violation of effect free programming?

### ISSUE THREAD

**Issue by Waynelearn** [nus-cs2030/2223-s2#44]
# Summary
As per above. If I change the state of servers in the shop, do I have to create a new shop accordingly?
Description of the question / problem here.

## Description:

Clarifications or details with respect to the issue you are facing

## Code Snippets

Attach code snippets (if any)

## Screenshots

Screenshots (if any)

**Comment by e0959170** [comment]
not sure if i'm understanding your question properly but any attribute in a class is required to be private final for submission so i think you would need to make a new shop object either way :)

**Comment by oofpouf** [comment]
Do you mean changing the the value of the instance variable of server? If so, as mentioned by the comment above, it does violate effect free programming. Codecrunch will require you to set your instance as private final and a new object has to be created.

**Comment by 0liverkong** [comment]
you should not need to have to create new shops. rethink what data is needed to be shared between servers, customers and the shop, and what data need not be initialised as part of Shop's private final attributes in Shop()

**Comment by danqiye1** [comment]
Just to clarify, effect-free just means that for the lifetime of single object (such as your `Shop` object), it’s internal variable should not mutate. That is why all your variables are supposed to be `private final`. As an example, if you have this method that tries to change the `Shop`:

```java
Shop updateServer(Shop shop) {
    shop.server = new Server();
    return shop;
}
```
Your old shop’s server will be mysteriously updated if you invoke this method without visibility of what the method does internally:
```java
Shop oldShop = new Shop(new Server());
Shop newShop = updateShop(oldShop);
// At this point newShop has the updated server, but oldShop also have the updated server. So there is a side effect on oldShop. 
```
The `final` modifier ensures that the compiler help you prevent such a mutation. With the `final` modifier, you write it like this:
```
Shop updateServer(Shop shop) {
    return new Shop(new Server);
}
```

If you use this `updateServer` method, we are creating a new object in heap, and thus you oldShop should still have the previous server while the returned newShop gets a new server. 

In summary, effect-free does not mean you do not update the internal state of `Shop`, just that you should update the state variables through creating a new `Shop` object rather than mutating the state variable of the old `Shop` object.
